### PR TITLE
mysql_user: prevent password getting set for existing users on on_cre…

### LINKED ISSUE #334: mysql.user tries to set password on second run with update_password = "on_create"
Hello

I did not find an issue describing my problem, tested witch community_mysql  3.1.2

##### SUMMARY
create user in mysql (8) with password validation configured:
password_history = 5
second run tries to update password (with the same value) which results in
 "msg": "(3638, \"Cannot use these credentials for 'bobby@localhost' because they contradict the password history policy\")"}
see below for ansible code used

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
mysql.user

##### ANSIBLE VERSION
```
2.9.27
```

##### COLLECTION VERSION
```
3.1.2
```

##### OS / ENVIRONMENT
Ubuntu 20.04 LTS on both sides

##### STEPS TO REPRODUCE
```yaml
---
- hosts: all
  tasks:
    - name: Create MySQL users
      become: yes
      community.mysql.mysql_user:
        plugin: caching_sha2_password
        name: "{{ item.name }}"
        plugin_auth_string: "{{ item.password }}"
        priv: "{{ item.priv }}"
        host: "localhost"
        update_password: "on_create"
        state: present
      with_items:
        - name: bobby
          password: ...
          priv: "sakila.*:SELECT"
```
and a smal .ini file describing which hosts to use

##### EXPECTED RESULTS
If update_password is set to "on_create" I would assume no alter user ... is issued at all, as long the user exists.

I'm not sure if it is possible to analyse authentication_string from mysql.user with the current password in the yaml
code in plugins/module_utils/user.py says (line 267)
            if plugin_auth_string and current_plugin[1] != plugin_auth_string:
                # this case can cause more updates than expected,
                # as plugin can hash auth_string in any way it wants
                # and there's no way to figure it out for
                # a check, so I prefer to update more often than never
                update = True
that it isn't, which will cause the same problem if update_password is set to always. 
But at least if update_password is set to on_create this code should not run at all.

##### ACTUAL RESULTS
```
msg": "(3638, \"Cannot use these credentials for 'bobby@localhost' because they contradict the password history policy\")"}
```

Probably this also will fail on every other plugin also, but I only tested caching_sha2_password.

Beside the actual error if I don't use plugin_auth_string but password, I get the user created but with an empty authentication_string,
the documentation could be a bit more clear when to use which variable. (Or the code could could handle "password" differently depending on the plugin given, so one who makes the plugin variable don't need to code around this depending on which plugin is used)

Regards

### PR DESCRIPTION
**Pull request by hubiongithub**
…ate when plugin is used

##### SUMMARY

if update_password is set to on_create the code did not update "password" but update "plugin,plugin_auth_string,plugin_hash_string"

From my point of view "on_create" should never update a password related value.

This change has the side effect that a change of the plugin (lets say caching_sha2_password -> auth_pam (which requires no plugin_auth_string at all) will not happen.

Fixes #334 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

mysql_user

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by hubiongithub** [comment]
I canvnot view the result from this third-party-check status: failure
I could not get the tests working here behind an outgoing proxy as apt runs inside (the ansible-test -docker) containers and apt does not accept HTTP_PROXY environment and I would have to set these proxy parameter inside  /etc/apt/apt-conf.d/.00_proxy.conf in all containers which ansible-test generates and running apt add key ... in. And I don't know how to do this.

**Comment by Andersson007** [comment]
@hubiongithub thanks for the PR!
I'll try to take a look at the changes and your comment at the beginning of next week.

Could you please also add a [changelog gragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment)?

File: changelogs/fragments/334-mysql_user_fix_logic_on_oncreate.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+bugfixes:
+  - "mysql_user - fix logicwhen update_password is set to on_create for users using plugin* arguments (https://github.com/ansible-collections/community.mysql/issues/334)."
+
+
```
Comment (Andersson007) [inline changelogs/fragments/334-mysql_user_fix_logic_on_oncreate.yml]:
```suggestion
  - "mysql_user - fix logic when ``update_password`` is set to ``on_create`` for users using ``plugin*`` arguments (https://github.com/ansible-collections/community.mysql/issues/334)."
```

**Comment by Andersson007** [comment]
@hubiongithub sorry for the late reply, busy days.., could you please rebase the PR and fix the changelog fragment? About rebasing, here's the [guide](https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html). As it's for ansible core, it uses the `devel` branch. Use `main` instead. Thanks!

@betanummeric @rsicart @Jorge-Rodriguez do you have experience in using these arguments? I'd be great to hear opinions of folks who use this feature

**Comment by betanummeric** [comment]
I have not used the plugin, plugin_auth_string, plugin_hash_string arguments yet, but I might start soon, for which I would need this PR too. Regardless of my potential usecase, this PR looks good to me. Some integration tests could be added.

**Comment by hubiongithub** [comment]
Rebase of PR to main: done
Format fixing of changelog fragment: done and pushed

File: changelogs/fragments/334-mysql_user_fix_logic_on_oncreate.yml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+bugfixes:
+  - "mysql_user - fix logicwhen ``update_password`` is set to ``on_create`` for users using ``plugin*`` arguments (https://github.com/ansible-collections/community.mysql/issues/334)."
```
Comment (Andersson007) [inline changelogs/fragments/334-mysql_user_fix_logic_on_oncreate.yml]:
```suggestion
  - "mysql_user - fix logic when ``update_password`` is set to ``on_create`` for users using ``plugin*`` arguments (https://github.com/ansible-collections/community.mysql/issues/334)."
```

File: plugins/modules/mysql_user.py [hunk 3]
```diff
@@ -456,8 +456,13 @@ def main():
                                             priv, append_privs, subtract_privs, tls_requires, module)
                 else:
                     changed, msg = user_mod(cursor, user, host, host_all, None, encrypted,
+<<<<<<< HEAD
                                             plugin, plugin_hash_string, plugin_auth_string,
                                             priv, append_privs, subtract_privs, tls_requires, module)
+=======
+                                            None, None, None,
+                                            priv, append_privs, tls_requires, module)
+>>>>>>> mysql_user: prevent password getting set for existing users on on_create when plugin is used
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
                                            None, None, None,
                                            priv, append_privs, subtract_privs, tls_requires, module)
```
You need to fix the conflict here, i.e. add `subtract_privs` to arguments in your version

**Comment by Andersson007** [comment]
@hubiongithub Could you also please clarify the documentation for corresponding arguments according to the change? (i would also suggest elaborating a bit on details in the changelog fragment's entry)

File: plugins/modules/mysql_user.py (hunk 3 above)
Comment (hubiongithub) [reply plugins/modules/mysql_user.py]: Oh, missed that, fixed.

File: changelogs/fragments/334-mysql_user_fix_logic_on_oncreate.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+bugfixes:
+  - "mysql_user - fix logic when ``update_password`` is set to ``on_create`` for users using ``plugin*`` arguments (https://github.com/ansible-collections/community.mysql/issues/334)."
+    - ``on_create`` sets ``password`` to None for old mysql_native_authentication but not for authentiation methods which uses the ``plugin*`` arguments
+    - this PR changes this so ``on_create`` also exchange ``plugin``, ``plugin_hash_string``, ``plugin_auth_string`` to None in the list of arguments to change
```
Comment (Andersson007) [inline changelogs/fragments/334-mysql_user_fix_logic_on_oncreate.yml]:
```suggestion
  - "mysql_user - fix logic when ``update_password`` is set to ``on_create`` for users using ``plugin*`` arguments. The ``on_create`` value sets ``password`` to None for old mysql_native_authentication but not for authentiation methods which uses the ``plugin*`` arguments. This fix changes this so ``on_create`` also exchange ``plugin``, ``plugin_hash_string``, ``plugin_auth_string`` to None in the list of arguments to change (https://github.com/ansible-collections/community.mysql/issues/334)."
```
Thanks! A small formatting change suggestion to make our changelog entries consistent.

There's also the `DOCUMENTATION` block in the module's file, could you please describe what you wrote in the fragment in the corresponding option descriptions?

File: plugins/modules/mysql_user.py [hunk 5]
```diff
@@ -22,7 +22,7 @@
     required: true
   password:
     description:
-      - Set the user's password.
+      - Set the user's password. Only for `mysql_native_password` authentication. For other authentication plugins see the combination of plugin, plugin_hash_string, plugin_auth_string.
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
      - Set the user's password. Only for C(mysql_native_password) authentication. For other authentication plugins see the combination of I(plugin), I(plugin_hash_string), I(plugin_auth_string).
```
Ansible has special format macroses for things like option names, etc. :)

File: plugins/modules/mysql_user.py [hunk 6]
```diff
@@ -115,8 +115,8 @@
     default: no
   update_password:
     description:
-      - C(always) will update passwords if they differ.
-      - C(on_create) will only set the password for newly created users.
+      - C(always) will update passwords if they differ. This affects password and the combinationi of plugin, plugin_hash_string, plugin_auth_string
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
      - C(always) will update passwords if they differ. This affects I(password) and the combination of I(plugin), I(plugin_hash_string), I(plugin_auth_string).
```

File: plugins/modules/mysql_user.py [hunk 7]
```diff
@@ -115,8 +115,8 @@
...
+      - C(on_create) will only set the password or the combination of plugin, plugin_hash_string, plugin_auth_string for newly created users.
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
      - C(on_create) will only set the password or the combination of plugin, plugin_hash_string, plugin_auth_string for newly created users.
```

**Review by Andersson007 — COMMENTED** [review]
you can just click `commit suggestion`

File: changelogs/fragments/334-mysql_user_fix_logic_on_oncreate.yml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - "mysql_user - fix logic when ``update_password`` is set to ``on_create`` for users using ``plugin*`` arguments (https://github.com/ansible-collections/community.mysql/issues/334).. The ``on_create`` sets ``password`` to None for old mysql_native_authentication but not for authentiation methods which uses the ``plugin*`` arguments. This PR changes this so ``on_create`` also exchange ``plugin``, ``plugin_hash_string``, ``plugin_auth_string`` to None in the list of arguments to change"
```
Comment (Andersson007) [inline changelogs/fragments/334-mysql_user_fix_logic_on_oncreate.yml]:
```suggestion
  - "mysql_user - fix logic when ``update_password`` is set to ``on_create`` for users using ``plugin*`` arguments (https://github.com/ansible-collections/community.mysql/issues/334). The ``on_create`` sets ``password`` to None for old mysql_native_authentication but not for authentiation methods which uses the ``plugin*`` arguments. This PR changes this so ``on_create`` also exchange ``plugin``, ``plugin_hash_string``, ``plugin_auth_string`` to None in the list of arguments to change"
```

File: plugins/modules/mysql_user.py [hunk 9]
```diff
@@ -22,7 +22,7 @@
     required: true
   password:
     description:
-      - Set the user's password.
+      - Set the user's password. Only for C(mysql_native_password) authentication. For other authentication plugins see the combination of I(plugin), I(plugin_hash_string), I(plugin_auth_string).
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
      - Set the user's password. Only for C(mysql_native_password) authentication.
        For other authentication plugins see the combination of I(plugin), I(plugin_hash_string), I(plugin_auth_string).
```
it should fix `ERROR: plugins/modules/mysql_user.py:25:161: E501: line too long (195 > 160 characters)` sanity error

File: plugins/modules/mysql_user.py [hunk 10]
```diff
@@ -22,7 +22,8 @@
     required: true
   password:
     description:
-      - Set the user's password.
+      - Set the user's password. Only for C(mysql_native_password) authentication.
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
      - Set the user's password. Only for C(mysql_native_password) authentication.
```

**Review by Andersson007 — COMMENTED** [review]
another linting issue

File: plugins/modules/mysql_user.py (hunk 10 above)
Comment (hubiongithub) [reply plugins/modules/mysql_user.py]: This trailing spaces is a real pain :-/ In the Email you can't even see this one. fixed.

Comment (Andersson007) [reply plugins/modules/mysql_user.py]: @hubiongithub yep:) BTW for future contributions, you can run sanity and other tests locally before pushing. See https://docs.ansible.com/ansible/devel/community/create_pr_quick_start.html (Prepare you environment and Test your changes chapters). Can save time.

**Review by Andersson007 — APPROVED** [review]

**Comment by Andersson007** [comment]
@hubiongithub thanks for the contribution! The community will be happy to see more from you!
@betanummeric thanks for reviewing!

**Comment by Andersson007** [comment]
@hubiongithub there are conflicts which need to be solved, so the automatic backporting is not possible. Would you like to backport your changes manually to stable-2 and stable-1 branches? See this [guide](https://docs.ansible.com/ansible/devel/community/development_process.html#backporting-merged-prs-in-ansible-base).
If you have no time, please let me know, i'll do it myself, np.

**Comment by hubiongithub** [comment]
I followed the steps from the patchback instructions above, I hope I didn't missed anything.

**Comment by Andersson007** [comment]
@hubiongithub i replied in https://github.com/ansible-collections/community.mysql/pull/379#issuecomment-1141944549

**Comment by Andersson007** [comment]
@hubiongithub what i usually do https://github.com/ansible-collections/community.mysql/pull/379#issuecomment-1141959272

**Pull request merged**